// @malagu/cli/lib/package/package-protocol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var npm_registry_1 = require("./npm-registry");
var ApplicationPackageOptions = /** @class */ (function () {
    function ApplicationPackageOptions() {
    }
    return ApplicationPackageOptions;
}());
exports.ApplicationPackageOptions = ApplicationPackageOptions;
function customizer(objValue, srcValue) {
    if (Array.isArray(objValue)) {
        return srcValue;
    }
}
exports.customizer = customizer;
var ComponentPackage = /** @class */ (function () {
    function ComponentPackage(raw) {
        this.raw = raw;
    }
    Object.defineProperty(ComponentPackage.prototype, "name", {
        get: function () {
            return this.raw.name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ComponentPackage.prototype, "version", {
        get: function () {
            return this.raw.version;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ComponentPackage.prototype, "description", {
        get: function () {
            return this.raw.description || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ComponentPackage.prototype, "malaguComponent", {
        get: function () {
            return this.raw.malaguComponent;
        },
        enumerable: true,
        configurable: true
    });
    ComponentPackage.prototype.getAuthor = function () {
        if (this.raw.publisher) {
            return this.raw.publisher.username;
        }
        if (typeof this.raw.author === 'string') {
            return this.raw.author;
        }
        if (this.raw.author && this.raw.author.name) {
            return this.raw.author.name;
        }
        if (!!this.raw.maintainers && this.raw.maintainers.length > 0) {
            return this.raw.maintainers[0].username;
        }
        return '';
    };
    return ComponentPackage;
}());
exports.ComponentPackage = ComponentPackage;
var RawComponentPackage;
(function (RawComponentPackage) {
    function is(pck) {
        return npm_registry_1.PublishedNodePackage.is(pck) && !!pck.keywords && pck.keywords.indexOf('malagu-component') !== -1;
    }
    RawComponentPackage.is = is;
})(RawComponentPackage = exports.RawComponentPackage || (exports.RawComponentPackage = {}));
