// @malagu/cli/lib/init/init-manager.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var fs_extra_1 = require("fs-extra");
var inquirer = require('inquirer');
var templates_1 = require("./templates");
var child_process_1 = require("child_process");
var hook_executor_1 = require("../hook/hook-executor");
var context_1 = require("../context");
var ora = require("ora");
var axios_1 = require("axios");
var packager_1 = require("../packager");
var chalk = require('chalk');
inquirer.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'));
var SEARCH_TEMPLATE_REPO_URI = 'https://api.github.com/search/repositories?q=topic:malagu-template&sort=stars&order=desc';
var PLACEHOLD = '{{ templatePath }}';
var InitManager = /** @class */ (function () {
    function InitManager(context) {
        this.context = context;
    }
    InitManager.prototype.output = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.selectTemplate()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.checkOutputDir()];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.doOutput()];
                    case 3:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.render = function () {
        return __awaiter(this, void 0, void 0, function () {
            var packageJsonPath, packageJson;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        packageJsonPath = path_1.resolve(this.outputDir, 'package.json');
                        return [4 /*yield*/, fs_extra_1.readJSON(packageJsonPath, { encoding: 'utf8' })];
                    case 1:
                        packageJson = _a.sent();
                        packageJson.name = this.context.name;
                        return [4 /*yield*/, fs_extra_1.writeJSON(packageJsonPath, packageJson, { encoding: 'utf8', spaces: 2 })];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.install = function () {
        return __awaiter(this, void 0, void 0, function () {
            var ctx, packagerId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getCliContext()];
                    case 1:
                        ctx = _a.sent();
                        packagerId = ctx.pkg.rootConfig.malagu.packager;
                        return [4 /*yield*/, packager_1.getPackager(packagerId).install(this.outputDir, {})];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.executeHooks = function () {
        return __awaiter(this, void 0, void 0, function () {
            var outputDir, _a, _b, _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        outputDir = this.outputDir;
                        process.chdir(outputDir);
                        _b = (_a = new hook_executor_1.HookExecutor()).executeInitHooks;
                        _d = (_c = context_1.HookContext).create;
                        return [4 /*yield*/, this.getCliContext()];
                    case 1: return [4 /*yield*/, _d.apply(_c, [_e.sent()])];
                    case 2: return [4 /*yield*/, _b.apply(_a, [_e.sent()])];
                    case 3:
                        _e.sent();
                        console.log(chalk(templateObject_1 || (templateObject_1 = __makeTemplateObject(["{bold.green Success!} Initialized \"", "\" example in {bold.blue ", "}."], ["{bold.green Success!} Initialized \"", "\" example in {bold.blue ", "}."])), this.name, outputDir));
                        process.exit(0);
                        return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.getCliContext = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!!this.cliContext) return [3 /*break*/, 2];
                        _a = this;
                        return [4 /*yield*/, context_1.CliContext.create(this.context.program, [], this.outputDir)];
                    case 1:
                        _a.cliContext = _b.sent();
                        this.cliContext.name = this.context.name;
                        this.cliContext.outputDir = this.context.outputDir;
                        _b.label = 2;
                    case 2: return [2 /*return*/, this.cliContext];
                }
            });
        });
    };
    Object.defineProperty(InitManager.prototype, "outputDir", {
        get: function () {
            return path_1.resolve(process.cwd(), this.context.outputDir, this.context.name);
        },
        enumerable: true,
        configurable: true
    });
    InitManager.prototype.checkOutputDir = function () {
        return __awaiter(this, void 0, void 0, function () {
            var answers;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!fs_extra_1.existsSync(this.outputDir)) return [3 /*break*/, 2];
                        return [4 /*yield*/, inquirer.prompt([{
                                    name: 'overwrite',
                                    type: 'confirm',
                                    message: 'App already exists, overwrite the app'
                                }])];
                    case 1:
                        answers = _a.sent();
                        if (!answers.overwrite) {
                            process.exit(-1);
                        }
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.toOfficialTemplate = function (name, location) {
        return { name: name + " " + chalk.italic.gray('Official'), value: { location: location, name: name } };
    };
    InitManager.prototype.toThirdPartyTemplate = function (item) {
        return { name: item.name + " " + chalk.italic.gray(item.stargazers_count + '⭑'), value: { location: item.clone_url, name: item.name } };
    };
    InitManager.prototype.selectTemplate = function () {
        return __awaiter(this, void 0, void 0, function () {
            var answers;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, inquirer.prompt([{
                                name: 'item',
                                type: 'autocomplete',
                                message: 'Select a template to init',
                                source: function (answersSoFar, input) { return __awaiter(_this, void 0, void 0, function () {
                                    var spinner, options, officialTemplates, items, thirdPartyTemplates, error_1;
                                    var _this = this;
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0:
                                                if (!!this.source) return [3 /*break*/, 5];
                                                spinner = ora({ text: 'loading...', discardStdin: false }).start();
                                                options = {
                                                    url: SEARCH_TEMPLATE_REPO_URI,
                                                    timeout: 5000,
                                                    headers: {
                                                        'User-Agent': 'Malagu CLI'
                                                    }
                                                };
                                                officialTemplates = Object.keys(templates_1.templates).map(function (key) { return _this.toOfficialTemplate(key, templates_1.templates[key]); });
                                                _a.label = 1;
                                            case 1:
                                                _a.trys.push([1, 3, 4, 5]);
                                                return [4 /*yield*/, axios_1.default.request(options)];
                                            case 2:
                                                items = (_a.sent()).items;
                                                thirdPartyTemplates = items.map(function (item) { return _this.toThirdPartyTemplate(item); });
                                                this.source = __spread(officialTemplates, thirdPartyTemplates);
                                                return [3 /*break*/, 5];
                                            case 3:
                                                error_1 = _a.sent();
                                                this.source = officialTemplates;
                                                return [2 /*return*/, this.source];
                                            case 4:
                                                spinner.stop();
                                                return [7 /*endfinally*/];
                                            case 5: return [2 /*return*/, this.source.filter(function (item) { return !input || item.name.toLowerCase().includes(input.toLowerCase()); })];
                                        }
                                    });
                                }); }
                            }])];
                    case 1:
                        answers = _a.sent();
                        this.name = answers.item.name;
                        this.context.name = this.context.name || answers.item.name;
                        this.location = answers.item.location;
                        return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.isLocalTemplate = function () {
        return !this.location.startsWith('http');
    };
    Object.defineProperty(InitManager.prototype, "realLocation", {
        get: function () {
            return this.location.replace(PLACEHOLD, path_1.resolve(__dirname, '..', '..', 'templates'));
        },
        enumerable: true,
        configurable: true
    });
    InitManager.prototype.doOutput = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.isLocalTemplate()) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.outputLocalTemplate()];
                    case 1:
                        _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        this.outputRemoteTempate();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.outputLocalTemplate = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, fs_extra_1.copy(this.realLocation, this.outputDir)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    InitManager.prototype.outputRemoteTempate = function () {
        child_process_1.spawnSync('git', ['clone', '--depth=1', this.location, this.outputDir], { stdio: 'inherit' });
    };
    return InitManager;
}());
exports.InitManager = InitManager;
var templateObject_1;
