// @malagu/mvc/lib/node/annotation/method.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../constants");
function Get(options) {
    return exports.Method('get', options);
}
exports.Get = Get;
function Post(options) {
    return exports.Method('post', options);
}
exports.Post = Post;
function Put(options) {
    return exports.Method('put', options);
}
exports.Put = Put;
function Patch(options) {
    return exports.Method('patch', options);
}
exports.Patch = Patch;
function Head(options) {
    return exports.Method('head', options);
}
exports.Head = Head;
function Delete(options) {
    return exports.Method('delete', options);
}
exports.Delete = Delete;
function Options(options) {
    return exports.Method('options', options);
}
exports.Options = Options;
exports.Method = function (method, options) {
    if (options === void 0) { options = ''; }
    return function (target, key, descriptor) {
        var metadata = { options: options, method: method, target: target, key: key, descriptor: descriptor };
        var metadataList = [];
        if (!Reflect.hasOwnMetadata(constants_1.METADATA_KEY.controllerMethod, target.constructor)) {
            Reflect.defineMetadata(constants_1.METADATA_KEY.controllerMethod, metadataList, target.constructor);
        }
        else {
            metadataList = Reflect.getOwnMetadata(constants_1.METADATA_KEY.controllerMethod, target.constructor);
        }
        metadataList.push(metadata);
    };
};
