// @malagu/cli/lib/webpack/config/base-config-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../../constants");
var TerserPlugin = require('terser-webpack-plugin');
var nodeExternals = require('webpack-node-externals');
var merge = require("webpack-merge");
var path = require("path");
var BaseConfigFactory = /** @class */ (function () {
    function BaseConfigFactory() {
    }
    BaseConfigFactory.prototype.create = function (config, context, target) {
        var dev = context.dev, pkg = context.pkg;
        var webpackMode = dev ? 'development' : 'production';
        var baseConfig = {
            name: target,
            mode: webpackMode,
            optimization: {
                minimize: !dev,
                minimizer: [
                    new TerserPlugin({
                        parallel: true,
                        terserOptions: {
                            keep_classnames: true,
                            keep_fnames: true
                        },
                        extractComments: false
                    })
                ]
            },
            devtool: dev ? 'eval-cheap-source-map' : undefined,
            stats: 'errors-only',
            resolve: {
                extensions: ['.tsx', '.ts', '.js']
            },
            module: {
                rules: [
                    {
                        test: /\.js$/,
                        enforce: 'pre',
                        use: [{ loader: 'source-map-loader' }],
                        exclude: /jsonc-parser/
                    },
                    {
                        test: /\.tsx?$/,
                        use: [{
                                loader: 'ts-loader',
                                options: {
                                    transpileOnly: true,
                                    experimentalWatchApi: true
                                },
                            }],
                        exclude: /node_modules/
                    }
                ]
            }
        };
        if (constants_1.BACKEND_TARGET === target) {
            var whitelist = pkg.componentPackages.map(function (cp) { return new RegExp(cp.name); });
            return merge(baseConfig, {
                target: 'node',
                externals: dev ? [nodeExternals({
                        whitelist: whitelist,
                        modulesDir: path.resolve(pkg.projectPath, '../node_modules')
                    }), nodeExternals({
                        whitelist: whitelist
                    }), nodeExternals({
                        whitelist: whitelist,
                        modulesDir: path.resolve(pkg.projectPath, '../../node_modules')
                    })] : undefined,
                node: {
                    __dirname: false,
                    __filename: false
                },
                devtool: 'source-map',
            });
        }
        else {
            return merge(baseConfig, {
                target: 'web',
                node: {
                    fs: 'empty',
                    child_process: 'empty',
                    net: 'empty',
                    crypto: 'empty'
                },
                performance: {
                    hints: false
                },
                stats: 'errors-only',
                module: {
                    rules: [
                        {
                            test: /worker-main\.js$/,
                            loader: 'worker-loader',
                            options: {
                                name: 'worker-ext.[hash].js'
                            }
                        },
                        {
                            test: /\.css$/,
                            exclude: /\.useable\.css$/,
                            use: [
                                'style-loader',
                                'css-loader'
                            ]
                        },
                        {
                            test: /\.useable\.css$/,
                            use: [
                                'style-loader/useable',
                                'css-loader'
                            ]
                        },
                        {
                            test: /\.(jpg|png|gif)$/,
                            loader: 'file-loader',
                            options: {
                                name: '[hash].[ext]',
                            }
                        },
                        {
                            test: /source-map-support/,
                            loader: 'ignore-loader'
                        },
                        {
                            test: /\.woff(2)?(\\?v=[0-9]\.[0-9]\.[0-9])?$/,
                            loader: 'url-loader',
                            options: {
                                limit: 10000,
                                mimetype: 'application/font-woff'
                            }
                        },
                        {
                            test: /\.wasm$/,
                            loader: 'file-loader',
                            type: 'javascript/auto',
                        },
                        {
                            test: /\.plist$/,
                            loader: 'file-loader',
                        }
                    ]
                }
            });
        }
    };
    BaseConfigFactory.prototype.support = function (context, target) {
        return true;
    };
    return BaseConfigFactory;
}());
exports.BaseConfigFactory = BaseConfigFactory;
