// @malagu/cli/lib/webpack/config/der-server-config-factory.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../../constants");
var utils_1 = require("../utils");
var DevServerConfigFactory = /** @class */ (function () {
    function DevServerConfigFactory() {
    }
    DevServerConfigFactory.prototype.create = function (config, context, target) {
        var pkg = context.pkg, port = context.port, open = context.open;
        var realPort = utils_1.getPort(pkg, target, port);
        var baseDevServerConfig = {
            stats: 'errors-only',
            watchOptions: {
                ignored: /node_modules/
            },
            port: realPort,
            open: open,
            contentBase: false
        };
        if (constants_1.BACKEND_TARGET === target) {
            return {
                devServer: __assign(__assign({}, baseDevServerConfig), { writeToDisk: true })
            };
        }
        else {
            var server = utils_1.getMalaguConfig(pkg, target).server || {};
            if (server.endpoint) {
                server.endpoint = server.endpoint.replace('{port}', realPort);
            }
            return {
                devServer: __assign({}, baseDevServerConfig)
            };
        }
    };
    DevServerConfigFactory.prototype.support = function (context, target) {
        return true;
    };
    return DevServerConfigFactory;
}());
exports.DevServerConfigFactory = DevServerConfigFactory;
