// @malagu/core/lib/browser/application/frontend-application.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
var shell_1 = require("../shell");
var frontend_application_state_1 = require("./frontend-application-state");
var browser_1 = require("../browser");
var common_1 = require("../../common");
var FrontendApplication = /** @class */ (function (_super) {
    __extends(FrontendApplication, _super);
    function FrontendApplication() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(FrontendApplication.prototype, "shell", {
        get: function () {
            return this._shell;
        },
        enumerable: true,
        configurable: true
    });
    FrontendApplication.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var host;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.doStart()];
                    case 1:
                        _a.sent();
                        this.stateService.state = 'started';
                        return [4 /*yield*/, this.getHost()];
                    case 2:
                        host = _a.sent();
                        this._shell.attach(host);
                        return [4 /*yield*/, new Promise(function (resolve) { return requestAnimationFrame(function () { return resolve(); }); })];
                    case 3:
                        _a.sent();
                        this.stateService.state = 'attached_shell';
                        return [4 /*yield*/, this.revealShell(host)];
                    case 4:
                        _a.sent();
                        this.registerEventListeners();
                        this.stateService.state = 'ready';
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Return a promise to the host element to which the application shell is attached.
     */
    FrontendApplication.prototype.getHost = function () {
        if (document.body) {
            return Promise.resolve(document.body);
        }
        return new Promise(function (resolve) {
            return window.addEventListener('load', function () { return resolve(document.body); }, { once: true });
        });
    };
    /**
     * Return an HTML element that indicates the startup phase, e.g. with an animation or a splash screen.
     */
    FrontendApplication.prototype.getStartupIndicator = function (host) {
        var startupElements = host.getElementsByClassName('malagu-preload');
        return startupElements.length === 0 ? undefined : startupElements[0];
    };
    /**
     * Register global event listeners.
     */
    FrontendApplication.prototype.registerEventListeners = function () {
        var _this = this;
        window.addEventListener('unload', function () {
            _this.stateService.state = 'closing_window';
            _this.doStop();
        });
    };
    /**
     * If a startup indicator is present, it is first hidden with the `malagu-hidden` CSS class and then
     * removed after a while. The delay until removal is taken from the CSS transition duration.
     */
    FrontendApplication.prototype.revealShell = function (host) {
        var startupElem = this.getStartupIndicator(host);
        if (startupElem) {
            return new Promise(function (resolve) {
                window.requestAnimationFrame(function () {
                    startupElem.classList.add('malagu-hidden');
                    var preloadStyle = window.getComputedStyle(startupElem);
                    var transitionDuration = browser_1.parseCssTime(preloadStyle.transitionDuration, 0);
                    window.setTimeout(function () {
                        var parent = startupElem.parentElement;
                        if (parent) {
                            parent.removeChild(startupElem);
                        }
                        resolve();
                    }, transitionDuration);
                });
            });
        }
        else {
            return Promise.resolve();
        }
    };
    FrontendApplication.prototype.doStart = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, lifecycle, _loop_1, this_1, _c, _d, lifecycle, e_1_1;
            var e_2, _e, e_1, _f;
            var _this = this;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        try {
                            for (_a = __values(this.lifecycles), _b = _a.next(); !_b.done; _b = _a.next()) {
                                lifecycle = _b.value;
                                if (lifecycle.initialize) {
                                    try {
                                        lifecycle.initialize();
                                    }
                                    catch (error) {
                                        this.logger.error('Could not initialize lifecycle', error);
                                    }
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_b && !_b.done && (_e = _a.return)) _e.call(_a);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                        _loop_1 = function (lifecycle) {
                            var error_1;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!lifecycle.onStart) return [3 /*break*/, 4];
                                        _a.label = 1;
                                    case 1:
                                        _a.trys.push([1, 3, , 4]);
                                        return [4 /*yield*/, this_1.measure(lifecycle.constructor.name + '.onStart', function () { return lifecycle.onStart(_this); })];
                                    case 2:
                                        _a.sent();
                                        return [3 /*break*/, 4];
                                    case 3:
                                        error_1 = _a.sent();
                                        this_1.logger.error('Could not start lifecycle', error_1);
                                        return [3 /*break*/, 4];
                                    case 4: return [2 /*return*/];
                                }
                            });
                        };
                        this_1 = this;
                        _g.label = 1;
                    case 1:
                        _g.trys.push([1, 6, 7, 8]);
                        _c = __values(this.lifecycles), _d = _c.next();
                        _g.label = 2;
                    case 2:
                        if (!!_d.done) return [3 /*break*/, 5];
                        lifecycle = _d.value;
                        return [5 /*yield**/, _loop_1(lifecycle)];
                    case 3:
                        _g.sent();
                        _g.label = 4;
                    case 4:
                        _d = _c.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_1_1 = _g.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (_d && !_d.done && (_f = _c.return)) _f.call(_c);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    FrontendApplication.prototype.measure = function (name, fn) {
        return __awaiter(this, void 0, void 0, function () {
            var startMark, endMark, result, _a, _b, item;
            var e_3, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        startMark = name + '-start';
                        endMark = name + '-end';
                        performance.mark(startMark);
                        return [4 /*yield*/, fn()];
                    case 1:
                        result = _d.sent();
                        performance.mark(endMark);
                        performance.measure(name, startMark, endMark);
                        try {
                            for (_a = __values(performance.getEntriesByName(name)), _b = _a.next(); !_b.done; _b = _a.next()) {
                                item = _b.value;
                                if (item.duration > 100) {
                                    console.warn(item.name + ' is slow, took: ' + item.duration + ' ms');
                                }
                                else {
                                    console.debug(item.name + ' took ' + item.duration + ' ms');
                                }
                            }
                        }
                        catch (e_3_1) { e_3 = { error: e_3_1 }; }
                        finally {
                            try {
                                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                            }
                            finally { if (e_3) throw e_3.error; }
                        }
                        performance.clearMeasures(name);
                        return [2 /*return*/, result];
                }
            });
        });
    };
    __decorate([
        common_1.Autowired(shell_1.ApplicationShell),
        __metadata("design:type", Object)
    ], FrontendApplication.prototype, "_shell", void 0);
    __decorate([
        common_1.Autowired(common_1.ApplicationStateService),
        __metadata("design:type", frontend_application_state_1.FrontendApplicationStateService)
    ], FrontendApplication.prototype, "stateService", void 0);
    FrontendApplication = __decorate([
        common_1.Component(common_1.Application)
    ], FrontendApplication);
    return FrontendApplication;
}(common_1.AbstractApplication));
exports.FrontendApplication = FrontendApplication;
